shared/common-adapters/markdown.js

```diff
@@ -0,0 +1,118 @@
+// @flow
+
+import Text from './text'
+import React from 'react'
+import {List} from 'immutable'
+
+import type {Props} from './markdown'
+import type {PropsOf} from '../constants/types/more'
+
+// Order matters, since we want to match the longer ticks first
+const openToClosePair = {
+  '```': '```',
+  '`': '`',
+  '*': '*',
+  '_': '_',
+  '~': '~',
+}
+
+type TagInfo<C> = {Component: Class<C>, props: PropsOf<C>}
+
+const initialOpenToTag = {
+  '`': {Component: Text, props: {type: 'CodeSnippet', style: {}}},
+  '```': {Component: Text, props: {type: 'CodeSnippetBlock'}},
+  '*': {Component: Text, props: {type: 'BodySemibold'}},
+  '_': {Component: Text, props: {type: 'BodyInherit', style: {fontStyle: 'italic'}}},
+  '~': {Component: Text, props: {type: 'BodyInherit', style: {textDecoration: 'line-through'}}},
+}
+
+const openToNextOpenToTag = {
+  '`': {},
+  '```': {},
+  '*': initialOpenToTag,
+  '_': initialOpenToTag,
+  '~': initialOpenToTag,
+}
+
+const plainStringTag = {Component: Text, props: {type: 'Body'}}
+
+type TagMeta = {
+  componentInfo: {Component: ReactClass<*>, props: Object},
+  textSoFar: string,
+  elementsSoFar: List<React$Element<*> | string>,
+  openToTag: {[key: string]: TagInfo<*>},
+  closingTag: ?string,
+}
+
+const initalTagMeta: TagMeta = {
+  componentInfo: plainStringTag,
+  textSoFar: '',
+  elementsSoFar: new List(),
+  openToTag: initialOpenToTag,
+  closingTag: null,
+}
+
+type TagStack = List<TagMeta>
+
+function matchWithMark (text: string): ?{matchingMark: string, restText: string} {
+  const matchingMark = Object.keys(openToClosePair).find(mark => text.indexOf(mark) === 0)
```

Comment:
:sheep: In the failure case the `indexOf` here will do a lot of extra work traversing to the end of the string. You could probably make this faster by generating a `Regexp` using all valid `openToClosePair` and testing it here.